### DISCUSSION TITLE
SettingsExpander and SettingsExpanderItem Binding

### DISCUSSION THREAD

**Discussion by alex6dj** [amwx/FluentAvalonia#291] [Q&A] [answered]
Im having a bad time with the new preview5 thanks to the new ItemsControl refactoring https://github.com/amwx/FluentAvalonia/issues/288, but I know it is a good change. Well Im not really good with Styling stuff, so following @amwx suggestion https://github.com/amwx/FluentAvalonia/issues/288#issuecomment-1421557509 I ended with an styling to fix the generation of SettingsExpanderItem inside an SettingsExpander with Binding Items, after some refactoring.

```xml
    <UserControl.Styles>
        <Style Selector="controls|SettingsExpanderItem">

            <Setter Property="Content">
                <Template>
                    <fileFilter:FileFilterContentControl />
                </Template>
            </Setter>

            <Setter Property="Footer">
                <Template>
                    <fileFilter:FileFilterButtonsControl />
                </Template>
            </Setter>

        </Style>
    </UserControl.Styles>
```

and I ended wth this:
![Captura de pantalla (59)](https://user-images.githubusercontent.com/27690651/217557040-94174882-7d7b-4470-a7e4-455145ec9e5a.png)

The Footer is generated correctly but not the Content. If copying the content template to the Footer just to test if the are any design problem it gets generated correctly (in the Footer), but no look with the Content property.

So, what other thing needs to be done to fix this problem.

**Comment by amwx** [comment, marked as answer]
I don't have enough info on what you're trying to do here, but I think what you really want is to use the `ContentTemplate` on `SettingsExpanderItem` to build your view, and let the content come in from your ViewModel. SettingsExpanderItem is already a ContentControl, so it works like any other with binding. The content should be automatically set as part of container generation, but if not, `<Setter Property="Content" Value="{Binding }" />` should make it work.

**Reply by alex6dj** [reply]
First, sorry for not be more explicit. I'm going to mark your last comment as the answer, but if you have the time to make another suggestion I'd appreciate it. Here are some details.

As example `Supported file extensions`, it looks this way in preview4:
![Captura de pantalla (64)](https://user-images.githubusercontent.com/27690651/217692404-449059ab-3b54-4c21-9d5b-71c9daf433ed.png)

So `SupportedFileExtensionSettingItem` contains a `ReadOnlyObservableCollection` property named `SupportedExtensions`. And SupportedExtensions contains a bunch of `ExtensionProxy`.

```csharp
public sealed class ExtensionProxy : ReactiveObject
{ 
    public ExtensionProxy(string extension)
    {
        ...
    }

    public string Extension { get; set; }

    public bool HasChanged { get; }

    public bool Editing { get; private set; }

    public ReactiveCommand<Unit, Unit> ToggleEditing { get; }
}
```

`SettingsExpander`  is contructed around `SupportedFileExtensionSettingItem`, and each `SettingsExpanderItem` inside the SettingsExpander  is represented by an `ExtensionProxy`. So Footer should contain the commands (Edit from ExtensionProxy and Delete from SupportedFileExtensionSettingItem). And Content should contain a graphical representation of the Extension property in this case.

I have two UserControl, first for Content and Second for Footer. Following your last comment I have a working sample now using Styles for Footer and DataTemplates for Content:
```xml
<UserControl xmlns="https://github.com/avaloniaui" x:DataType="settings:SupportedFileExtensionSettingItem">
    <UserControl.Styles>
        /// This fill the SettingsExpanderItem Footer. I'm not sure if this is the correct way to do it
        <Style Selector="controls|SettingsExpanderItem">
            <Setter Property="Footer">
                <Template>
                    <fileFilter:FileFilterButtonsControl />
                </Template>
            </Setter>
        </Style>
    </UserControl.Styles>

    <UserControl.DataTemplates>
        /// This fill the SettingsExpanderItem Content
        <DataTemplate x:DataType="settings:ExtensionProxy">
            <fileFilter:FileFilterContentControl />
        </DataTemplate>
    </UserControl.DataTemplates>

    <StackPanel Margin="{StaticResource SettingsViewMargin}" Spacing="4">

        <controls:SettingsExpander Header="" Items="{CompiledBinding  SupportedExtensions}">

            <controls:SettingsExpander.IconSource>
                <controls:PathIconSource Data="{DynamicResource FilterIcon}" />
            </controls:SettingsExpander.IconSource>

            <controls:SettingsExpander.Footer>
                <Button Command="{CompiledBinding Add}" Height="32">
                    <controls:FAPathIcon Data="{DynamicResource NewFileIcon}" />
                </Button>
            </controls:SettingsExpander.Footer>

        </controls:SettingsExpander>

    </StackPanel>

</UserControl>
```

Hey, you got here. Well thanks for your time.

**Reply by amwx** [reply]
Based on my understanding, yes you're doing it correctly.
